io/module_write_netcdf.F90

```diff
@@ -555,8 +557,73 @@ subroutine add_dim(ncid, dim_name, dimid, grid, rc)
     end if
 
     call get_grid_attr(grid, dim_name, ncid, dim_varid, rc)
+    
+    ! if iau_offset set, update time units attribute.
+    if ( trim(dim_name) == "time" .and. iau_offset > 0) then
+! get time units just written to file
+       ncerr = nf90_get_att(ncid, dim_varid, 'units', time_units); NC_ERR_STOP(ncerr)
+! idate: year,month,day,hour,minute,second
+       idate = get_idate_from_time_units(time_units)
+! idat: year,month,day,time zone,hour,minute,second,millsecond
+       idat = 0
+       idat(1:3) = idate(1:3)
+       idat(5:7) = idate(4:6)
+! rinc: days,hours,minutes,seconds,milliseconds
+       rinc = 0; rinc(2) = iau_offset
+       call w3movdat(rinc,idat,jdat)
+! update idate using iau_offset
+       idate(1:3)=jdat(1:3)
+       idate(4:6)=jdat(5:7)
+       time_units  = get_time_units_from_idate(idate)
+! rewrite time units attribute.
+       ncerr = nf90_put_att(ncid, dim_varid, 'units', trim(time_units)); NC_ERR_STOP(ncerr)
+    endif
```

Comment:
Can this manipulation of 'units' attribute be done outside of this subroutine? At the place where data and attributes are added to ESMF_FieldBundle. This routine (write_netcdf) should be as generic as possible. It shouldn't "know" anything about iau_offset.

Reply:
Ideally yes, but I'm afraid that model may use this ESMF units attribute internally.  Maybe Jun can confirm whether it's used or not?

Reply:
I looked at the code again. Write grid component actually uses IO_BASETIME ( ESMF_Time) and IO_CURRTIMEDIFF (ESMF_TimeIntervalESMF time type) in its internal state to specify output initial date and the forecast hours related to initial output date.  These two variable can be changed from forecast state due to various reasons, and in current code write grid component already adjusts the two time variables if iau_offset>0. Instead getting the initial output time from write grid, I think you can pass these two time variables into write_netcdf subroutine.

Reply:
The "time:units" attribute is set in module_fcst_grid_comp.F90. I see iau_offset is available in that module. Can "time:units" attribute be created with correct value in that module and passed to write_netcdf routine via FieldBundle. Instead of writing incorrect value, reading it back, parsing individual time components, adjusting based on iau_offset, constructing new "hours since ..." string and finally overwriting the one already written.
Passing IO_BASETIME and IO_CURRTIMEDIFF instead of iau_offset and doing the same or similar manipulation is equally bad.

Reply:
"Time:units" in fcst_grid_comp is to control the forecast integration and it is correct to provide time information from forecast point of view (starting integration from 2019103121 for 2019110100 cycle, but we decided only to take forecast results for 2019110100 history output). In my point of view "Time:units" on forecast grid comp should not be changed, because it is write grid comp decided what to output with the data. In the case of running with IAU, it's 2019110100 is the corresponding f00. That is why we use the two time variables  IO_BASETIME and IO_CURRTIMEDIFF owned by write grid comp to have consistent time stamp for all the tasks done on write grid comp(output files, inline_post etc).

Reply:
The "time:units" is a string attribute added to the export state for the purpose of passing that information (and many other) to the routine that writes that state to a (netcdf) file (convention="NetCDF"). I do not see how it can "control the forecast". Where exactly is "time:units" attribute used, except in the write_netcdf routine?  If it's used then that's also not the best way of controlling the forecast. We use various esmf objects (clocks, time, time intervals and alarms) for that, not string attributes.

Reply:
Sorry, I should say "time:units" on forecast grid component shows the time stamp on forecast integration, no matter what downstream routine(s) will use it, we should not change it. Write grid comp should be allowed to make decision on what to output.

Reply:
I implemented Dusan's suggestion and it seems to work.